src/renderer/entities/network/lib/provider/WrappedScProvider.ts

```diff
@@ -0,0 +1,19 @@
+import { ScProvider } from '@polkadot/rpc-provider';
+
+import { GET_METADATA_METHOD } from '../common/constants';
+import { HexString } from '@renderer/domain/shared-kernel';
+
+export const createWrappedScProvider = (getMetadata: () => Promise<HexString | undefined>) => {
+  class WrappedScProvider extends ScProvider {
+    async send(method: string, params: unknown[]): Promise<any> {
+      if (method === GET_METADATA_METHOD && params.length === 0) {
+        const metadata = await getMetadata();
+        if (metadata) return metadata;
+      }
+
+      return super.send(method, params);
+    }
+  }
+
+  return WrappedScProvider;
+};
```

Comment:
Why did we decide to step away from solution described in the research docs?

Initital plan was to have a single wrapped provider, lets say `MetadataCachingProvider`, which will accept an instance of any other `Provider` in its constructor and will delegate all calls except GET_METADATA_METHOD to this instance.

This way, we dont have to create two similar implementations (`createWrappedScProvider ` and `createWrappedWsProvider `) but rather have a single wrapper that can be combined with any other `Provider` implementation